### ISSUE TITLE
Integration of window handle Eltako FHF

### ISSUE THREAD

**Issue by LutzDi** [grimmpp/home-assistant-eltako#32]
Hi @grimmpp,

besides the fact that I still have issues in getting the v.1.2.3 up and running, I started with the integration of the window handle Eltako FHF from Hoppe. I saw that the EEP F6-10-00 is supported and in my understanding, I have to set this as a sensor and not as a binary_sensor because the handle has three states OPEN, CLOSE, and TILT.

If I set the config up with the following code

```
  sensor:
    - id: "00-21-20-D1"
      name: "Küche Fenster rechts"
      eep: "F6-10-00"
```
the handle is shown as a sensor, but nothing happened, no state change. The debug logs show the following entries:

```
2023-12-30 16:35:00.825 DEBUG (MainThread) [eltako] Received message: <ESP2Message '0b 05 e0 00 00 00 00 21 20 d1 20'>
2023-12-30 16:35:00.829 ERROR (SyncWorker_7) [homeassistant.util.logging] Exception in _message_received_callback when dispatching 'eltako.receive_message': (<ESP2Message '0b 05 e0 00 00 00 00 21 20 d1 20'>,)
Traceback (most recent call last):
  File "/config/custom_components/eltako/device.py", line 69, in _message_received_callback
    self.value_changed(msg)
  File "/config/custom_components/eltako/sensor.py", line 406, in value_changed
    if decoded.learn_button != 1:
       ^^^^^^^^^^^^^^^^^^^^
AttributeError: 'F6_10_00' object has no attribute 'learn_button'

2023-12-30 16:35:01.211 DEBUG (MainThread) [eltako] Received message: <ESP2Message '0b 05 d0 00 00 00 00 21 20 d1 20'>
2023-12-30 16:35:01.216 ERROR (SyncWorker_9) [homeassistant.util.logging] Exception in _message_received_callback when dispatching 'eltako.receive_message': (<ESP2Message '0b 05 d0 00 00 00 00 21 20 d1 20'>,)
Traceback (most recent call last):
  File "/config/custom_components/eltako/device.py", line 69, in _message_received_callback
    self.value_changed(msg)
  File "/config/custom_components/eltako/sensor.py", line 406, in value_changed
    if decoded.learn_button != 1:
       ^^^^^^^^^^^^^^^^^^^^
AttributeError: 'F6_10_00' object has no attribute 'learn_button'

2023-12-30 16:35:01.705 DEBUG (MainThread) [eltako] Received message: <ESP2Message '0b 05 c0 00 00 00 00 21 20 d1 20'>
2023-12-30 16:35:01.710 ERROR (SyncWorker_3) [homeassistant.util.logging] Exception in _message_received_callback when dispatching 'eltako.receive_message': (<ESP2Message '0b 05 c0 00 00 00 00 21 20 d1 20'>,)
Traceback (most recent call last):
  File "/config/custom_components/eltako/device.py", line 69, in _message_received_callback
    self.value_changed(msg)
  File "/config/custom_components/eltako/sensor.py", line 406, in value_changed
    if decoded.learn_button != 1:
       ^^^^^^^^^^^^^^^^^^^^
AttributeError: 'F6_10_00' object has no attribute 'learn_button'

2023-12-30 16:35:01.991 DEBUG (MainThread) [eltako] Received message: <ESP2Message '0b 05 f0 00 00 00 00 21 20 d1 20'>
2023-12-30 16:35:01.996 ERROR (SyncWorker_6) [homeassistant.util.logging] Exception in _message_received_callback when dispatching 'eltako.receive_message': (<ESP2Message '0b 05 f0 00 00 00 00 21 20 d1 20'>,)
Traceback (most recent call last):
  File "/config/custom_components/eltako/device.py", line 69, in _message_received_callback
    self.value_changed(msg)
  File "/config/custom_components/eltako/sensor.py", line 406, in value_changed
    if decoded.learn_button != 1:
       ^^^^^^^^^^^^^^^^^^^^
AttributeError: 'F6_10_00' object has no attribute 'learn_button'
```

Can you help me with what I don't see and/or understand? Could bit be that this is also a result of the fact that I didn't get the new config running correctly? I appreciate your support in this...

Thx and cheers
Lutz

**Comment by LutzDi** [comment]
Just for conformation: After I got my config for v1.2.3 sorted out and up and running I can still confirm that the issue with the window handle as sensor is still the same, also the error message stays the same... I would appreciate some help form you guys..

Thx and cheers
Lutz

**Comment by grimmpp** [comment]
Hello @LutzDi,

I haven't looked into how to split up HA configuration and what should be considered also from integration logic point of view.

Yes, I can imagine that you need to completely recreate all entities. Through the new gateway elements I needed to create a combined Id for entities so that they can be address clearly. There are people who made actually an migration of their old entities to the new ones to keep there data history. Unfortunately, I don't really know how it works, I just know they changed the entity ids in the database.

Regarding the **Window Handles** let me know if you need support. I have no Hoppe handles but I know the issue that it is hard to tell HA what you want, especially because window contacts are binary sensors. 
F6-10-00 is supported in both binary_sensor and sensor but binary_sensor has only state open or closed. :-) 
It should be supported like you can see [here](https://github.com/grimmpp/home-assistant-eltako/blob/8fb0010e08661808384fccfe3b000e37c2e4942e/custom_components/eltako/sensor.py#L442). 

Please, let me know if it works or when you need help.

All the best and good luck

**Comment by LutzDi** [comment]
Hi @grimmpp,

thx for your answer... Yes, I think starting from scratch is the easiest way and I don't care about the data history of my components, not needed for those kinds of devices...

Coming back to the window handle: Unfortunately, I'm not experienced enough to "translate" the information from the sensor.py you mentioned to any meaningful configuration within HA. I can imagine that creating a template binary_sensor could help so that open/tilt leads to the status OPEN for the template bin_sen and close leads to the status "CLOSE". But of course, then I lose the status tilt and to be honest, for me, it's interesting to know if the window is really open or only tilted... That makes a difference if I'm away to decide whether must get nervous or not :).

So I decided to approach with the standard Enocean integration for the window handles - and what to say: works like a charm. I use your integration with the FGW14USB and the Enocean integration with the USB300 stick. It took me 10 min to see the status of the window handle correctly. Perhaps you already know [but here is the link to the configuration for a window handle with the standard Enocean integration of HA.](https://www.home-assistant.io/integrations/enocean#window-handle).

Last but not least leads me this experience to the question: If it's possible to get everything correct in Ha concerning the window handle with the standard Enocean integration, what is the missing puzzle piece for your integration to get the right status of the window handle. And this is the point where I have to beg your support because I don't have the knowledge to do it on my own. I'm interested in reducing the number of integrations to reduce risks and raise the stability and performance of the system and at the moment I only need a solution for the window handle with your integration, everything else works. Let me know how I can support you with testing or information, this is what I can do with my limited knowledge...

I'm looking forward to your answer...

Cheers
Lutz

**Comment by grimmpp** [comment]
Hello @LutzDi,

that should be an easy task. Yes, the integrations differs because Eltako is based on the older ESP2 protocol and USB300 uses ESP3 which uses different telegrams for the usb connection. (Still on my list to support both but has lower prio right now. The bigger thing I'm working on will be a new serial connector under the hood which is able to automatically recover so that you do not need to manually restart HA for reconnecting.) 
In the integration the data of the window handle which is check and evaluated is the same. It could be that the learn_mode is always activated and then the integration just ignores it. Some sensors just have learn_mode always on.
Can you send me three example telegrams of your window handle containing the different states? Then I will troubleshoot it. Do you know how to switch on the log messages?
I will extend the tests first to also ensure the stability for further changes. (Similar like I did it for the [weather station tests](https://github.com/grimmpp/home-assistant-eltako/blob/main/tests/test_sensor.py))

**Comment by LutzDi** [comment]
> Can you send me three example telegrams of your window handle containing the different states? Then I will troubleshoot it. Do you know how to switch on the log messages?

Of course, it will be my pleasure... Could you lead me to the right way to collect the logs you need? Do you mean collecting the telegrams with DolphinView, or can it be done with HA debug logging? I appreciate your help... thx a lot.

**Comment by grimmpp** [comment]
Would be good if you could give me the plain byte array.
Most probably you won't see this in the HA logs. How logging in HA works is describe e.g. [here](https://github.com/grimmpp/home-assistant-eltako/blob/main/docs/rocker_switch/readme.md).
For proper monitoring I use [WinEtel](https://www.eltako.com/en/download_file/enocean-tool-winetel/)

**Comment by LutzDi** [comment]
Hi @grimmpp,

unfortunately, WinETel and me, we won't be friends, I didn't get it working, I only get errors that the "Serial Telegram is incorrectly received" :)... But I used DolphinView which does the same in my understanding. I've attached the results from DW for u, sorry for the screenshots, but I didn't find a solution to export it to a file. I've made the test from CLOSE (6 o'clock position) to OPEN (3 o'clock position) to tilt (12 o'clock position) and back to open and to close. I named the files accordingly 

<img width="632" alt="Window Handle close-to-open" src="https://github.com/grimmpp/home-assistant-eltako/assets/139575137/e8696b9b-6b75-49eb-9e31-41db612c697a">
<img width="637" alt="Window Handle open-to-close" src="https://github.com/grimmpp/home-assistant-eltako/assets/139575137/609a3d9e-f235-4969-98b1-780b0e6dfee9">
<img width="618" alt="Window Handle open-to-tilt" src="https://github.com/grimmpp/home-assistant-eltako/assets/139575137/bddb1e59-2bf1-4197-bebd-00e6d501d6c5">
<img width="657" alt="Window Handle tilt-to-open" src="https://github.com/grimmpp/home-assistant-eltako/assets/139575137/01dbd840-6ef2-4b91-8792-3149fe6c2293">

just a question for my better understanding: These here are the entries in the debug log from HA for the same actions:
```
2024-01-02 08:48:42.461 DEBUG (MainThread) [eltako] [Gateway] [Id: 1] Received message: <ESP2Message '0b 05 e0 00 00 00 00 21 20 d1 20'>
2024-01-02 08:48:44.203 DEBUG (MainThread) [eltako] [Gateway] [Id: 1] Received message: <ESP2Message '0b 05 d0 00 00 00 00 21 20 d1 20'>
2024-01-02 08:48:45.960 DEBUG (MainThread) [eltako] [Gateway] [Id: 1] Received message: <ESP2Message '0b 05 c0 00 00 00 00 21 20 d1 20'>
2024-01-02 08:48:47.622 DEBUG (MainThread) [eltako] [Gateway] [Id: 1] Received message: <ESP2Message '0b 05 f0 00 00 00 00 21 20 d1 20'>
```
Would you find also the needed information in these entries? It's just for me to learn something...:)

I hope you find the right and needed information, let me know if something is missing...

Thx and cheers
Lutz

**Comment by grimmpp** [comment]
Hello @LutzDi,

that's great. Actually the log lines would have been enough but every interesting to see the DolphinView results I haven't been able to get it working. :-) Do you use USB300 for it in conjunction?

The messages look like described in the [official standard](https://www.enocean-alliance.org/wp-content/uploads/2017/05/EnOcean_Equipment_Profiles_EEP_v2.6.7_public.pdf) on page 25 and 26. 

[Values in the code](https://github.com/grimmpp/home-assistant-eltako/blob/8fb0010e08661808384fccfe3b000e37c2e4942e/custom_components/eltako/sensor.py#L442) do not match at all. 

I can make it work. Give me a few days time.

**Comment by LutzDi** [comment]
>  Do you use USB300 for it in conjunction?

Yes, I use a normal Win10 device with my USB300, the FTDI-Driver will be installed with DolphinView automatically or can be downloaded > [HERE](https://ftdichip.com/drivers/vcp-drivers/): But if I remember right, FTDI is also part of Win10 by default. Then you get a COM-Port mapped to the USB300 in Windows and everything else was straightforward.. Just connect and have fun... :).

And yes, if you show me the code lines I understand the point, also without coding experience :)... Do u need every position mentioned in the official documentation? Or do you take the values from the documentation as they are?

And of course, no prob, take your time... I'm ready for testing if you need something...

Have a great day.
Lutz

**Comment by grimmpp** [comment]
Was quicker than I thought. 

Test are here: https://github.com/grimmpp/home-assistant-eltako/blob/f58bfd38aa48dc02f564182aaf2e2700797801ab/tests/test_sensor.py#L80

Changes are pushed to `main` branch.

Can you test it in real?

**Comment by LutzDi** [comment]
Hi @grimmpp,

mille grazie... Could you give me a hin how to test it in the best way? I mean, I don't know exactly what I should to with the test_sensor.py. Should I replace the original sensor.py with this for a test? Thx for a short explanation how to use the test_sensor.py...

cheers
Lutz

**Comment by grimmpp** [comment]
Actually, I didn't mean the tests. Can you test it in real in your automation with your window handles? Just reinstall the integration and take main as version

**Comment by LutzDi** [comment]
Thx, @grimmpp.

Perhaps a fault on my side, but it's a bit weird:

- I had the integration loaded in the prior version and the window handle was configured. It was shown under Eltako devices
- If I remove the configuration for this window handle from the config, the device stays and isn't deleted.
  - 1st question: How can I get rid of old devices that do not exist anymore? At the moment, it seems, that the integration doesn't double-check during boot for the actual config and removes non-existing devices during this check. I've tested this also with a switch (binary_sensor), this also stays in the HA config under devices also if I remove it from the configuration. HA is saying that the corresponding entity isn't provided anymore by Eltako and can be deleted. Yes, I can delete this entity, but the device itself stays with no entity and I don't see any way to delete the device. Do you have an idea?
- Coming back to my test: I think I stuck with the registration of devices and entities because also with the new main branch version HA tells me that the device is configured and connected to the FGW14-USB, but the corresponding entity "Window Handle" isn't provided by the integration Eltako. I think, I stuck here...

What I have done after I struggeled:

- Switch back to a working backup with the "old" version 1.3.0. Deleted the existing Eltako integration and that HACS entry
- Rebooted HA
- Added Eltako in HACS loaded Eltako integration MAIN again
- RESULT: It seems that the files are still the old ones, no changes in mainfest.json and sensor.py like shown in github
- Added my complete configuration of all Eltako devices again, but without the window handle
- RESULT: All works as expected
- Exchange manifest.json and sensor.py on file-level against the version from the new branch from github from yesterday
- Rebooted HA
- RESULT: All still woks like expected
- Added the window handle to the configuration
- Rebooted HA
- RESULT: Device window handle was added as well as the entity but the entity is shown as "not provided by integration Eltako". Here is the German error message, I think you understand German :): "Diese Entität wird nicht mehr von der Integration eltako bereitgestellt. Wenn die Entität nicht mehr verwendet wird, lösche sie in den Einstellungen."

So long story short: With the provided version it's not working. I also tried the 1.3.0 without the window handle optimization, but the entity of the window handle isn't provided by this version as well. I went back to V1.2.3 and tried the same and the behavior is the same. As it is like always: I'm pretty sure that the windows handle entity existed with V1.2.3 normally, but I can't swear on it.

For me it seems that there is something broke in registering device an/or entity of the window handle in the right way.

Any idea, what to do?

Cheers
Lutz

**Comment by grimmpp** [comment]
Hello @LutzDi,

1. It is wanted behavior of HA that it keeps the old entries for data history reasons. That is independent of the integration. I cannot delete entities.
2. Very strange that you cannot download main easily via HACS. I tried it our for myself and it is not working as well for me. I always get v1.3.0. You can also simply use my [install script](https://github.com/grimmpp/home-assistant-eltako/blob/main/docs/install-specific-version-or-branch.md).
3. There should be actually no change in the creation step of the sensor entity. Sounds very strange to me. Did you check the list of entities listed under the gateway? There you should see all the entities which were loaded. Even if the do not exist in real, they should be listed when they are in the configuration.
  If tested it with
  ```
      sensor:
      - id: ff-bc-c8-50
        eep: F6-10-00
        name: Window Handle
  ```
  and that is my result
  ![grafik](https://github.com/grimmpp/home-assistant-eltako/assets/12120359/312c51ab-0887-4115-8c14-d9d57590d6f7)
  I also don't see any error regarding this entry in the logs.
  I also managed to simulate the window handle with WinEtel and send the telegrams to HA. There was a little bug because HA was expecting numeric values instead of strings (open, closed, tilt). I fixed that. Now it should work.
  ![grafik](https://github.com/grimmpp/home-assistant-eltako/assets/12120359/659defa2-1236-48c8-81f4-2ffab6aec59b)
![grafik](https://github.com/grimmpp/home-assistant-eltako/assets/12120359/a2a671f0-6058-4907-80f6-723ee11f0d2b)

Changes are again pushed to main branch. Just use [install script](https://github.com/grimmpp/home-assistant-eltako/blob/main/docs/install-specific-version-or-branch.md) to test it. (You can leave out step 4 containing`git checkout`)

**Comment by LutzDi** [comment]
Good morning @grimmpp,

thx for all those explanations... I restarted my tests with the installation of the integration via the install-script. After that, the files are fine and the same as the main branch...

And, what to say: It works like a charm. The sensor is listed correctly as a device as well as the corresponding entity, with no error, and the status is shown correctly. CLOSE-OPEN-TILT, all as it should be... So the question about the window handle can be closed from my side...

The only point is the question of why the installation via HACS failed. But this is independent of the window handle integration. I think installing the integration via install-script did the trick... 

Thx a lot for your support.

cheers and happy weekend.
Lutz

**Comment by grimmpp** [comment]
Cool, then I'll create a release. That should solve the issue with HACS.
